File: pricing/RouteFinder.ts
import UniswapV2Pair from "./AMM.js";
import Route from "./Route.js";
import Token from "./Token.js";

class RouteFinder {
  // Finds optimal routes between tokens.
  pools: UniswapV2Pair[];

  // token.name -> neighbors
  graph: Map<string, Array<{ pool: UniswapV2Pair; token: Token }>>;

  constructor(pools: UniswapV2Pair[]) {
    this.pools = pools;
    this.graph = this.buildGraph();
  }

  buildGraph(): Map<string, Array<{ pool: UniswapV2Pair; token: Token }>> {
    const g = new Map<string, Array<{ pool: UniswapV2Pair; token: Token }>>();

    for (const p of this.pools) {
      if (!g.has(p.token0.name)) g.set(p.token0.name, []);
      if (!g.has(p.token1.name)) g.set(p.token1.name, []);

      g.get(p.token0.name)!.push({ pool: p, token: p.token1 });
      g.get(p.token1.name)!.push({ pool: p, token: p.token0 });
    }

    return g;
  }

  findAllRoutes(tokenIn: Token, tokenOut: Token, maxHops: number = 3): Route[] {
    /*
        Find all possible routes up to max_hops.
        */
    const routes: Route[] = [];

    const dfs = (
      current: Token,
      visited: Set<string>,
      poolsPath: UniswapV2Pair[],
      tokensPath: Token[],
      hopsLeft: number,
    ) => {
      if (hopsLeft < 0) return;

      if (current.equals(tokenOut)) {
        routes.push(new Route([...poolsPath], [...tokensPath]));
        return;
      }

      const neighbors = this.graph.get(current.name);
      if (!neighbors) return;

      for (const { pool, token } of neighbors) {
        if (visited.has(token.name)) continue;

        visited.add(token.name);

        poolsPath.push(pool);
        tokensPath.push(token);

        dfs(token, visited, poolsPath, tokensPath, hopsLeft - 1);

        poolsPath.pop();
        tokensPath.pop();

        visited.delete(token.name);
      }
    };

    dfs(tokenIn, new Set([tokenIn.name]), [], [tokenIn], maxHops);

    return routes;
  }
  convertToOutputToken(gasWei: bigint, tokenOut: Token): bigint {
    if (tokenOut.name === "ETH") return gasWei;

    const pool = this.pools.find(
      (p) =>
        (p.token0.name === "ETH" && p.token1.equals(tokenOut)) ||
        (p.token1.name === "ETH" && p.token0.equals(tokenOut)),
    );

    if (!pool) {
      throw new Error("No ETH→tokenOut pool for gas conversion");
    }

    const ethToken = new Token("ETH", 10n ** 18n);

    // simulate swap of gasWei ETH → tokenOut
    return pool.getAmountOut(
      Number(gasWei) / Number(ethToken.decimals),
      ethToken,
    );
  }
  findBestRoute(
    tokenIn: Token,
    tokenOut: Token,
    amountIn: number,
    gasPriceGwei: bigint,
    maxHops: number = 3,
  ): [Route, bigint] {
    /*
        Find route that maximizes NET output (after gas).
        Returns (best_route, net_output).
        */
    const routes = this.findAllRoutes(tokenIn, tokenOut, maxHops);

    let bestRoute: Route | undefined = undefined;
    let bestNetOutput = BigInt(0);

    routes.forEach((route) => {
      const grossOutput = route.getOutput(amountIn);
      const gasCost = route.estimateGas() * gasPriceGwei * BigInt(1000000000);
      const gasCostInOutputToken = this.convertToOutputToken(gasCost, tokenOut);
      const netOutput = grossOutput - gasCostInOutputToken;
      console.log(route);
      console.log(grossOutput - gasCostInOutputToken);
      if (netOutput > bestNetOutput) {
        bestNetOutput = netOutput;
        bestRoute = route;
      }
    });
    return [bestRoute!, bestNetOutput];
  }

  compareRoutes(
    tokenIn: Token,
    tokenOut: Token,
    amountIn: number,
    gasPriceGwei: bigint,
  ) {
    /*
        Compare all routes with detailed breakdown:
        {
            'route': Route,
            'grossOutput': int,
            'gasEstimate': int,
            'gasCost': int,
            'netOutput': int,
        }
        */
    const routes = this.findAllRoutes(tokenIn, tokenOut);

    const results = [];

    for (const route of routes) {
      const gross = route.getOutput(amountIn);
      const gasEstimate = route.estimateGas();
      const gasCostWei = gasEstimate * gasPriceGwei * 1_000_000_000n;

      const net = gross - gasCostWei;

      results.push({
        route,
        gross_output: gross,
        gas_estimate: gasEstimate,
        gas_cost: gasCostWei,
        net_output: net,
      });
    }

    return results.sort((a, b) => (a.net_output > b.net_output ? -1 : 1));
  }
}

export default RouteFinder;

File: pricing/route_finder.py
from core.base_types import Address
from typing import Optional
from pricing.AMM import UniswapV2Pair
from pricing.route import Route
from pricing.token import Token


class RouteFinder:
    """
    Finds optimal routes between tokens.
    """

    def __init__(self, pools: list[UniswapV2Pair]):
        self.pools: list[UniswapV2Pair] = pools
        self.graph: dict = self._build_graph()

    def _build_graph(self) -> dict:
        """
        Build adjacency graph: token → [(pool, other_token), ...]
        """
        g: dict[str, dict[UniswapV2Pair, Token]] = {}
        for p in self.pools:
            if g.get(p.token0.name) is None:
                g[p.token0.name] = {}
            if g.get(p.token1.name) is None:
                g[p.token1.name] = {}

            g.get(p.token0.name)[p] = p.token1
            g.get(p.token1.name)[p] = p.token0

        return g

    def find_all_routes(
        self, token_in: Token, token_out: Token, max_hops: int = 3
    ) -> list[Route]:
        """
        Find all possible routes up to max_hops.
        """
        routes: list[Route] = []

        def dfs(
            current: Token,
            visited: {str},
            pools_path: list[UniswapV2Pair],
            tokens_path: list[Token],
            hopsLeft: int,
        ):
            if hopsLeft < 0:
                return

            if current.__eq__(token_out):
                routes.append(Route(pools_path.copy(), tokens_path.copy()))
                return

            neighbors: dict[UniswapV2Pair, Token] = self.graph.get(current.name)

            if len(neighbors) == 0:
                return

            for pool, token in neighbors.items():
                if token.name in visited:
                    continue

                visited.add(token.name)

                pools_path.append(pool)
                tokens_path.append(token)

                dfs(token, visited, pools_path, tokens_path, hopsLeft - 1)

                pools_path.pop()
                tokens_path.pop()

                visited.remove(token.name)

        dfs(token_in, {token_in.name}, [], [token_in], max_hops)

        return routes

    def convert_to_output_token(self, gas_wei: int, token_out: Token) -> int:
        if token_out.address.checksum == "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2":
            return gas_wei
        ethToken = Token(
            "WETH",
            10**18,
            Address.from_string("0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2"),
        )
        routes = self.find_all_routes(ethToken, token_out, 5)
        minGas = routes[0].get_output(gas_wei)
        for i in range(0, len(routes)):
            out = routes[i].get_output(gas_wei)
            if out < minGas:
                minGas = out

        # simulate swap of gasWei WETH → tokenOut
        return minGas

    def find_best_route(
        self,
        token_in: Token,
        token_out: Token,
        amount_in: int,
        gas_price_gwei: int,
        max_hops: int = 3,
    ) -> tuple[Route, int]:
        """
        Find route that maximizes NET output (after gas).
        Returns (best_route, net_output).
        """
        routes = self.find_all_routes(token_in, token_out, max_hops)
        bestRoute: Optional[Route] = None
        bestNetOutput = 0
        for route in routes:
            grossOutput = route.get_output(amount_in)

            gasCost = route.estimate_gas() * gas_price_gwei * 1_000_000_000
            gasCostInOutputToken = self.convert_to_output_token(gasCost, token_out)
            netOutput = grossOutput - gasCostInOutputToken
            if netOutput > bestNetOutput:
                bestNetOutput = netOutput
                bestRoute = route

        return bestRoute, bestNetOutput

    # def compare_routes(
    #     self,
    #     token_in: Token,
    #     token_out: Token,
    #     amount_in: int,
    #     gas_price_gwei: int
    # ) -> list[dict]:
    #     """
    #     Compare all routes with detailed breakdown:
    #     {
    #         'route': Route,
    #         'gross_output': int,
    #         'gas_estimate': int,
    #         'gas_cost': int,
    #         'net_output': int,
    #     }
    #     """
